contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,479 @@
+    * Which allows discovery and health check of devices.
+    * Which allows hooking the runtime to make devices available in containers
+      and cleaning them up.
+2. Define a deployment mechanism for this new API.
+3. Define a versioning mechanism for this new API.
+
+# Non Objectives
+
+1. Handling heterogeneous nodes and topology related problems
+2. Collecting metrics is not part of this proposal. We will only solve
+   Health Check.
+
+# TLDR
+
+At their core, device plugins are simple gRPC servers that may run in a
+container deployed through the pod mechanism.
+
+These servers implement the gRPC interface defined later in this design
+document and once the device plugin makes itself known to kubelet, kubelet
+will interact with the device through three simple functions:
+  1. A `ListWatch` function for the kubelet to Discover the devices and
+     their properties as well as notify of any status change (device
+     became unhealthy).
+  2. An `Allocate` function which is called before creating a user container
+     consuming any exported devices
+
+![Process](device-plugin-overview.png)
+
+# Vendor story
+
+Kubernetes provides to vendors a mechanism called device plugins to:
+  * advertise devices.
+  * monitor devices (currently perform health checks).
+  * hook into the runtime to execute device specific instructions
+    (e.g: Clean GPU memory) and 
+    to take in order to make the device available in the container.
+
+```go
+service DevicePlugin {
+	// returns a stream of []Device
+	rpc ListWatch(Empty) returns (stream ListWatchResponse) {}
+	rpc Allocate(AllocateRequest) returns (AllocateResponse) {}
+}
+```
+
+The gRPC server that the device plugin must implement is expected to
+be advertised on a unix socket in a mounted hostPath (e.g:
+`/var/run/Kubernetes/vendor.sock`).
+
+Finally, to notify Kubelet of the existence of the device plugin,
+the vendor's device plugin will have to make a request to Kubelet's
+own gRPC server.
+Only then will kubelet start interacting with the vendor's device plugin
+through the gRPC apis.
+
+# End User story
+
+When setting up the cluster the admin knows what kind of devices are present
+on the different machines and therefore can select what devices to enable.
+
+The cluster admin knows his cluster has NVIDIA GPUs therefore he deploys
+the NVIDIA device plugin through:
+`kubectl create -f nvidia.io/device-plugin.yml`
+
+The device plugin lands on all the nodes of the cluster and if it detects that
+there are no GPUs it terminates (assuming `restart: OnFailure`). However, when
+there are GPUs it reports them to Kubelet and starts it's gRPC server to
+monitor devices and hook into the container creation process.
+
+Devices reported by Device Plugins are advertised as Extended resources of
+the shape `extensions.kubernetes.io/vendor-device`.
+E.g., Nvidia GPUs are advertised as nvidia-gpu
+
+
+Devices can be selected using the same process as for OIRs in the pod spec.
```

Comment:
it would be good if we state that consumption of devices implies minimum of Burstable QOS class or not.

i.e. will devices become a resource that is tracked by qos (like cpu or memory) or if its independent of that.